File: go/algorithms/sliding_window_counter.go
package algorithms

import (
	"math"
	"sync"
	"time"
)

type windowData struct {
	prevCount       int
	currCount       int
	prevWindowStart time.Time
	currWindowStart time.Time
}

type SlidingWindowCounter struct {
	mu         sync.RWMutex
	windows    map[string]*windowData
	limit      int
	windowSize time.Duration
}

func NewSlidingWindowCounter(limit int, windowSize time.Duration) *SlidingWindowCounter {
	return &SlidingWindowCounter{
		windows:    make(map[string]*windowData),
		limit:      limit,
		windowSize: windowSize,
	}
}

func (swc *SlidingWindowCounter) Allow(key string) bool {
	swc.mu.Lock()
	defer swc.mu.Unlock()

	now := time.Now()
	currentWindowStart := now.Truncate(swc.windowSize)

	data, exists := swc.windows[key]
	if !exists {
		data = &windowData{
			prevCount:       0,
			currCount:       0,
			prevWindowStart: currentWindowStart.Add(-swc.windowSize),
			currWindowStart: currentWindowStart,
		}
		swc.windows[key] = data
	}

	// Check if we need to slide the window
	if data.currWindowStart.Before(currentWindowStart) {
		data.prevCount = data.currCount
		data.prevWindowStart = data.currWindowStart
		data.currCount = 0
		data.currWindowStart = currentWindowStart
	}

	// Calculate the weighted count
	windowProgress := float64(now.Sub(currentWindowStart)) / float64(swc.windowSize)
	prevWindowWeight := 1.0 - windowProgress
	estimatedCount := float64(data.prevCount)*prevWindowWeight + float64(data.currCount)

	if estimatedCount < float64(swc.limit) {
		data.currCount++
		return true
	}

	return false
}

func (swc *SlidingWindowCounter) Reset(key string) {
	swc.mu.Lock()
	defer swc.mu.Unlock()
	delete(swc.windows, key)
}

func (swc *SlidingWindowCounter) GetStats(key string) RateLimiterStats {
	swc.mu.RLock()
	defer swc.mu.RUnlock()

	now := time.Now()
	currentWindowStart := now.Truncate(swc.windowSize)

	data, exists := swc.windows[key]
	if !exists || data.currWindowStart.Before(currentWindowStart) {
		return RateLimiterStats{
			Allowed:   true,
			Limit:     swc.limit,
			Remaining: swc.limit,
			ResetAt:   currentWindowStart.Add(swc.windowSize),
		}
	}

	// Calculate the weighted count
	windowProgress := float64(now.Sub(currentWindowStart)) / float64(swc.windowSize)
	prevWindowWeight := 1.0 - windowProgress
	estimatedCount := float64(data.prevCount)*prevWindowWeight + float64(data.currCount)
	remaining := int(math.Max(0, float64(swc.limit)-estimatedCount))

	return RateLimiterStats{
		Allowed:   estimatedCount < float64(swc.limit),
		Limit:     swc.limit,
		Remaining: remaining,
		ResetAt:   currentWindowStart.Add(swc.windowSize),
	}
}

File: typescript/src/algorithms/sliding-window-counter.ts
import { RateLimiter, RateLimiterStats } from './rate-limiter.interface';

interface WindowData {
  prevCount: number;
  currCount: number;
  prevWindowStart: number;
  currWindowStart: number;
}

export class SlidingWindowCounter implements RateLimiter {
  private windows: Map<string, WindowData> = new Map();
  private limit: number;
  private windowSize: number;

  constructor(limit: number, windowSizeMs: number) {
    this.limit = limit;
    this.windowSize = windowSizeMs;
  }

  async allow(key: string): Promise<boolean> {
    const now = Date.now();
    const currentWindowStart = Math.floor(now / this.windowSize) * this.windowSize;
    
    let data = this.windows.get(key);
    
    if (!data) {
      data = {
        prevCount: 0,
        currCount: 0,
        prevWindowStart: currentWindowStart - this.windowSize,
        currWindowStart: currentWindowStart
      };
      this.windows.set(key, data);
    }

    if (data.currWindowStart < currentWindowStart) {
      data.prevCount = data.currCount;
      data.prevWindowStart = data.currWindowStart;
      data.currCount = 0;
      data.currWindowStart = currentWindowStart;
    }

    const windowProgress = (now - currentWindowStart) / this.windowSize;
    const prevWindowWeight = 1 - windowProgress;
    const estimatedCount = data.prevCount * prevWindowWeight + data.currCount;

    if (estimatedCount < this.limit) {
      data.currCount++;
      return true;
    }

    return false;
  }

  async reset(key: string): Promise<void> {
    this.windows.delete(key);
  }

  async getStats(key: string): Promise<RateLimiterStats> {
    const now = Date.now();
    const currentWindowStart = Math.floor(now / this.windowSize) * this.windowSize;
    const data = this.windows.get(key);
    
    if (!data || data.currWindowStart < currentWindowStart) {
      return {
        allowed: true,
        limit: this.limit,
        remaining: this.limit,
        resetAt: new Date(currentWindowStart + this.windowSize)
      };
    }

    const windowProgress = (now - currentWindowStart) / this.windowSize;
    const prevWindowWeight = 1 - windowProgress;
    const estimatedCount = data.prevCount * prevWindowWeight + data.currCount;
    const remaining = Math.max(0, Math.floor(this.limit - estimatedCount));

    return {
      allowed: estimatedCount < this.limit,
      limit: this.limit,
      remaining,
      resetAt: new Date(currentWindowStart + this.windowSize)
    };
  }
}